standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java

```diff
@@ -3362,95 +3362,84 @@ private void updateStatsForTruncate(Map<String,String> props, EnvironmentContext
+              environmentContext);
+        }
+      }
     }
-    alterHandler.alterPartition(ms, wh, catName, dbName, tableName, null, partition,
-        environmentContext, this, validWriteIds);
   }
 
   private void alterTableStatsForTruncate(RawStore ms, String catName, String dbName,
-                                          String tableName, Table table, List<String> partNames,
-                                          String validWriteIds, long writeId) throws Exception {
-    if (partNames == null) {
-      if (0 != table.getPartitionKeysSize()) {
-        for (Partition partition : ms.getPartitions(catName, dbName, tableName, -1)) {
-          alterPartitionForTruncate(ms, catName, dbName, tableName, table, partition,
-              validWriteIds, writeId);
-        }
-      } else {
-        EnvironmentContext environmentContext = new EnvironmentContext();
-        updateStatsForTruncate(table.getParameters(), environmentContext);
-
-        boolean isReplicated = isDbReplicationTarget(ms.getDatabase(catName, dbName));
-        if (!transactionalListeners.isEmpty()) {
-          MetaStoreListenerNotifier.notifyEvent(transactionalListeners,
-              EventType.ALTER_TABLE,
-              new AlterTableEvent(table, table, true, true,
-                  writeId, this, isReplicated));
-        }
-
-        if (!listeners.isEmpty()) {
-          MetaStoreListenerNotifier.notifyEvent(listeners,
-              EventType.ALTER_TABLE,
-              new AlterTableEvent(table, table, true, true,
-                  writeId, this, isReplicated));
-        }
-
-        // TODO: this should actually pass thru and set writeId for txn stats.
-        if (writeId > 0) {
-          table.setWriteId(writeId);
-        }
-        alterHandler.alterTable(ms, wh, catName, dbName, tableName, table,
-            environmentContext, this, validWriteIds);
-      }
+      String tableName, Table table, List<Partition> partitionsList,
+      String validWriteIds, long writeId) throws Exception {
+    if (0 != table.getPartitionKeysSize()) {
+      alterPartitionsForTruncate(ms, catName, dbName, tableName, table, partitionsList,
+          validWriteIds, writeId);
     } else {
-      for (Partition partition : ms.getPartitionsByNames(catName, dbName, tableName, partNames)) {
-        alterPartitionForTruncate(ms, catName, dbName, tableName, table, partition,
-            validWriteIds, writeId);
+      EnvironmentContext environmentContext = new EnvironmentContext();
+      updateStatsForTruncate(table.getParameters(), environmentContext);
+      environmentContext.putToProperties(hive_metastoreConstants.IS_TRUNCATE_OP, Boolean.TRUE.toString());
+
+      // TODO: this should actually pass thru and set writeId for txn stats.
+      if (writeId > 0) {
+        table.setWriteId(writeId);
       }
+      alterHandler.alterTable(ms, wh, catName, dbName, tableName, table,
```

Comment:
nit: cloud we also call `ms.alterTable` directly for truncating the table? so we don't need to update the column statistics in alterHandler and using the `IS_TRUNCATE_OP` to flag the operation. Rest looks good to me.

Reply:
Done